### PR TITLE
SQL optimization for UI-focused bulk metric history API

### PR DESCRIPTION
**Pull request by dbczumar**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

This PR extends #7593 to introduce an optimized bulk metric fetch operation for SQL-based stores. The diff will shrink once #7593 is merged.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by dbczumar** [comment]
@mlflow-automation autoformat

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -1031,24 +1031,43 @@ def get_metric_history_bulk_handler():
...
+                        "timestamp": metric.timestamp,
+                        "step": metric.step,
+                        "run_id": run_id,
+                    }
+                    for metric in metrics_for_run
+                ]
+            )
+        return metrics_with_run_ids
+
+    if hasattr(store, "get_metric_history_bulk"):
```
Comment (apurva-koti) [inline mlflow/server/handlers.py:1060]: Nice!

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 2]
```diff
@@ -862,6 +862,70 @@ def get_metric_history(self, run_id, metric_key, max_results=None, page_token=No
             metrics = session.query(SqlMetric).filter_by(run_uuid=run_id, key=metric_key).all()
             return PagedList([metric.to_mlflow_entity() for metric in metrics], None)
 
+    class MetricWithRunId(Metric):
+        def __init__(self, run_id, metric: Metric):
```
Comment (apurva-koti) [inline mlflow/store/tracking/sqlalchemy_store.py:866]:
```suggestion
        def __init__(self, metric: Metric, run_id):
```

Comment (apurva-koti) [reply mlflow/store/tracking/sqlalchemy_store.py:866]: Nit, to reflect class name and information hierarchy

**Review by apurva-koti — APPROVED** [review]

File: mlflow/store/tracking/sqlalchemy_store.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/store/tracking/sqlalchemy_store.py]: Good point! Committed!

**Review by hubertzub-db — APPROVED** [review]
seems lgtm!

**Pull request merged**